### PR TITLE
Comprehensive VSM Guide

### PR DESCRIPTION
**Pull request by albertbori**
## Description

Added a comprehensive guide with placeholders for reference articles.

Note: The linked reference guides are not complete. Those will be done in a follow-up PR.

## Type of Change

- [ ] Bug Fix
- [ ] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [x] Documentation
- [ ] Other (please describe)

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/wayfair-incubator/vsm-ios/blob/main/CONTRIBUTING.md)
- [x] Existing issues have been referenced (where applicable)
- [x] I have verified this change is not present in other open pull requests
- [x] Functionality is documented
- [x] All code style checks pass
- [x] New code contribution is covered by automated tests
- [x] All new and existing tests pass

### CODE REVIEW

File: Sources/VSM/Documentation.docc/ComprehensiveGuide/ViewDefinition-UIKit.md [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    super.init(coder: coder)
+}
+
+// Encapsulated
+required init?(state: LoadUserProfileViewState, coder: NSCoder) {
+    let loaderModel = LoadUserProfileViewState.LoaderModel(userId: 1)
+    let state = .initialized(loaderModel)
+    container = .init(state: state)
+    super.init(coder: coder)
+}
```
Comment (granoff) [inline Sources/VSM/Documentation.docc/ComprehensiveGuide/ViewDefinition-UIKit.md:357]: Shouldn't this be:
```swift
required init?(userId: Int, coder: NSCoder) {
    let loaderModel = LoadUserProfileViewState.LoaderModel(userId: userId)
    let state = .initialized(loaderModel)
    container = .init(state: state)
    super.init(coder: coder)
}
```

**Review by granoff — APPROVED** [review]
This is amazing work. I've left some suggestions and some questions.

File: Sources/VSM/Documentation.docc/ComprehensiveGuide/UnitTesting.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        .expectSuccess()
+        .waitForExpectations(timeout: 10)
+}
+```
+
+> Tip: Be sure to use the ["Test Repeatedly"](https://www.avanderlee.com/debugging/flaky-tests-test-repetitions/) feature in Xcode to ensure that your Combine unit tests do not have flaky assertions.
+
+## Behavior-driven Development
+
+[Behavior-driven Development](https://en.wikipedia.org/wiki/Behavior-driven_development) (BDD) is an approach where writing intentionally-failing unit tests is part of requirements gathering process. VSM development shares some concepts with BDD in that you have to have a deep understanding of the requirements to build the shape of the feature before you can start implementing the view or the models. While BDD is not explicitly required for VSM, it may be a good option to consider adding to your process.
```
[the remark this answers is not in the document]
Comment (albertbori) [reply Sources/VSM/Documentation.docc/ComprehensiveGuide/UnitTesting.md:205]: Good news! They both do that! BDD is an extension of TDD. From the article I linked:
> It emerged from [test-driven development](https://en.wikipedia.org/wiki/Test-driven_development) (TDD). Behavior-driven development combines the general techniques and principles of TDD with ideas from [domain-driven design](https://en.wikipedia.org/wiki/Domain-driven_design) and [object-oriented analysis and design](https://en.wikipedia.org/wiki/Object-oriented_analysis_and_design) to provide software development and management teams with shared tools and a shared process to collaborate on software development.

The reason I chose BDD was that it aligns with the early requirements gathering exercise that is implicit in VSM development. Let me know if this doesn't convince you. We can definitely chat more about it. My mind is not fully set on it.

File: Sources/VSM/Documentation.docc/Reference/ModelActions.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Model Actions - VSM Reference
+
+A reference for choosing the best Action formats for your Models
+
+## Overview
+
+Actions are responsible for progressing the "State Journey" of the feature. They need to be flexible to account for any type of functionality. The VSM framework supports a handful of Action types within its ``StateContainer/observe(_:)-1uta3`` overloads. This article covers each one and gives examples of where they each are appropriately used.
```
[the remark this answers is not in the document]
Comment (albertbori) [reply Sources/VSM/Documentation.docc/Reference/ModelActions.md:7]: Xcode's DocC auto doc symbol link generator. That's how it disambiguates between overloads. It is really weird.

File: Sources/VSM/Documentation.docc/ComprehensiveGuide/DataDefinition.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        return Fail(error: error).eraseToAnyPublisher()
+    }
+    return URLSession.shared.dataTaskPublisher(for: request)
+        .tryMap(\.data)
+        .decode(type: UserData.self, decoder: JSONDecoder())
+        .map { userData -> UserData in
+            userDataSubject.value = .loaded(userData)
+            return userData
+        }
+        .eraseToAnyPublisher()
```
[the remark this answers is not in the document]
Comment (albertbori) [reply Sources/VSM/Documentation.docc/ComprehensiveGuide/DataDefinition.md:88]: Great point! I will simplify both the code and the explanation here.

File: Sources/VSM/Documentation.docc/ComprehensiveGuide/StateDefinition.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+
+Translating requirements into abstract types in code can be a bit challenging at first. By using the right process, it can quickly become second nature.
+
+### Extracting the Concepts from the Feature Requirements
+
+The best way to build features in VSM is to carefully read the requirements and translate those requirements into a flow chart. While this step is not required, it will help you more easily identify the states, data, and actions that a feature needs.
+
+Consider the following requirements.
+
+```gherkin
```
[the remark this answers is not in the document]
Comment (granoff) [reply Sources/VSM/Documentation.docc/ComprehensiveGuide/StateDefinition.md:68]: Ha! That's hillarious.

File: Sources/VSM/Documentation.docc/ComprehensiveGuide/VSMOverview.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+- State & model definitions are a simple and clear description of the feature requirements (for ease of maintenance)
+- Implementation code is easy to read
+- Passively encourages "Shifting Left" via Behavior-Driven Development
+
+### Cons
+
+- Inferring states, models, data, and actions from feature requirements can be challenging
+- Like most other architectures, hanging execution paths within actions are possible
+- Requires some data type translation for consumption by SwiftUI views
+
```
[the remark this answers is not in the document]
Comment (albertbori) [reply Sources/VSM/Documentation.docc/ComprehensiveGuide/VSMOverview.md:69]: Very true! The same could be said for any architecture. I'm trying to highlight cons that are unique to VSM, which set it apart from other architectures (both good and bad)

File: Sources/VSM/Documentation.docc/ComprehensiveGuide/ViewDefinition-UIKit.md [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+
+    override func viewDidLoad() {
+        stateSubscription = container.$state
+            .sink { [weak self] newState in
+                self?.render(newState)
+            }
+        super.viewDidLoad()
+    }
+
+    func render(state: VSMBasicExampleViewState) {
```
[the remark this answers is not in the document]
Comment (albertbori) [reply Sources/VSM/Documentation.docc/ComprehensiveGuide/ViewDefinition-UIKit.md:40]: Good catch!!

**Pull request merged**